File: sdks/python/perf_harness/drive/load.py
"""Bounded load plans: independent arrival rate and full-request concurrency."""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Literal


@dataclass(frozen=True)
class Stage:
    duration_s: float
    request_rate: float
    max_inflight: int
    kind: Literal["hold", "ramp", "warmup"] = "hold"
    name: str | None = None

    def __post_init__(self) -> None:
        if not math.isfinite(self.duration_s) or self.duration_s <= 0:
            raise ValueError("stage.duration_s must be finite and > 0")
        if math.isnan(self.request_rate) or self.request_rate < 0:
            raise ValueError("stage.request_rate must be >= 0 or inf")
        if (
            isinstance(self.max_inflight, bool)
            or not isinstance(self.max_inflight, int)
            or self.max_inflight < 1
        ):
            raise ValueError("stage.max_inflight must be an integer >= 1")
        if self.kind not in {"hold", "ramp", "warmup"}:
            raise ValueError("stage.kind must be hold, ramp or warmup")

    @property
    def level(self) -> float:
        return self.max_inflight if math.isinf(self.request_rate) else self.request_rate

    @property
    def label(self) -> str:
        return self.name or f"{self.kind}@{self.level:g}"


@dataclass(frozen=True)
class Warmup:
    """Double the request rate from 1 every step; zero disables warmup."""

    step_s: float = 5.0

    def __post_init__(self) -> None:
        if not math.isfinite(self.step_s) or self.step_s < 0:
            raise ValueError("warmup.step_s must be finite and >= 0")


@dataclass(frozen=True, kw_only=True)
class LoadPlan:
    request_rate: float
    max_inflight: int
    hold_s: float = 60.0
    warmup: Warmup = field(default_factory=Warmup)
    stages: tuple[Stage, ...] = ()
    arrival: Literal["constant", "poisson"] = "constant"
    seed: int = 0
    abort_on_error_rate: float | None = None
    breaker_min_n: int = 20
    cooldown_timeout_s: float = 180.0

    def __post_init__(self) -> None:
        Stage(self.hold_s, self.request_rate, self.max_inflight)
        if self.arrival not in {"constant", "poisson"}:
            raise ValueError("load.arrival must be constant or poisson")
        if not self.stages and self.request_rate <= 0:
            raise ValueError("load.request_rate must be > 0")
        if not math.isfinite(self.cooldown_timeout_s) or self.cooldown_timeout_s < 0:
            raise ValueError("load.cooldown_timeout_s must be finite and >= 0")
        if self.abort_on_error_rate is not None and not 0 < self.abort_on_error_rate <= 1:
            raise ValueError("load.abort_on_error_rate must be in (0, 1]")
        if (
            isinstance(self.breaker_min_n, bool)
            or not isinstance(self.breaker_min_n, int)
            or self.breaker_min_n < 1
        ):
            raise ValueError("load.breaker_min_n must be an integer >= 1")
        if any(math.isinf(s.request_rate) != self.saturated for s in self.stages):
            raise ValueError("stages cannot switch between finite request_rate and inf")

    @property
    def saturated(self) -> bool:
        return math.isinf(self.request_rate)

    @property
    def mode(self) -> str:
        return "concurrency" if self.saturated else "rate"

    @property
    def planned_stages(self) -> tuple[Stage, ...]:
        if self.stages:
            return self.stages
        stages = []
        rate = min(1.0, self.request_rate)
        if not self.saturated and self.warmup.step_s:
            while rate < self.request_rate:
                stages.append(Stage(self.warmup.step_s, rate, self.max_inflight, "warmup"))
                rate = min(rate * 2, self.request_rate)
        stages.append(Stage(self.hold_s, self.request_rate, self.max_inflight))
        return tuple(stages)

    @property
    def duration_s(self) -> float:
        """Planned upper bound; an early inflight limit shortens warmup."""
        return sum(s.duration_s for s in self.planned_stages)

    @property
    def peak_level(self) -> float:
        return max(s.level for s in self.planned_stages)

    @property
    def peak_inflight(self) -> int:
        return max(self.max_inflight, *(s.max_inflight for s in self.planned_stages))

    def target(self, elapsed_s: float) -> tuple[float, int]:
        rate, concurrency = self.request_rate, self.max_inflight
        clock = 0.0
        for stage in self.planned_stages:
            if elapsed_s < clock + stage.duration_s:
                if stage.kind != "ramp":
                    return stage.request_rate, stage.max_inflight
                fraction = max(0.0, elapsed_s - clock) / stage.duration_s
                # Infinity is a replenish policy, never an interpolated number.
                r = math.inf if self.saturated else rate + (stage.request_rate - rate) * fraction
                return r, max(
                    1, math.floor(concurrency + (stage.max_inflight - concurrency) * fraction)
                )
            clock += stage.duration_s
            rate, concurrency = stage.request_rate, stage.max_inflight
        return rate, concurrency

    def arrival_time(self, volume: float) -> float:
        """Invert integrated rate, so scheduler stalls never reset the arrival clock."""
        clock, previous = 0.0, self.request_rate
        for stage in self.planned_stages:
            start = previous if stage.kind == "ramp" else stage.request_rate
            slope = (stage.request_rate - start) / stage.duration_s
            mass = (start + stage.request_rate) * stage.duration_s / 2
            if volume < mass:
                if abs(slope) < 1e-12:
                    return clock + volume / start
                root = math.sqrt(max(0.0, start * start + 2 * slope * volume))
                return clock + (2 * volume / (start + root) if volume else 0.0)
            volume -= mass
            clock += stage.duration_s
            previous = stage.request_rate
        return math.inf

    def label(self) -> str:
        return f"{self.mode}/{self.peak_level:g}/c{self.peak_inflight}"

File: sdks/typescript/perf-harness/src/load.ts
export interface Stage {
  duration_s: number;
  request_rate: number;
  max_inflight: number;
  kind: "hold" | "ramp" | "warmup";
  name?: string;
}

export interface Warmup {
  step_s?: number;
}

export interface LoadPlan {
  request_rate: number;
  max_inflight: number;
  hold_s?: number;
  warmup?: Warmup;
  stages?: Stage[];
  arrival?: "constant" | "poisson";
  seed?: number;
  abort_on_error_rate?: number;
  breaker_min_n?: number;
  cooldown_timeout_s?: number;
}

export function stages(load: LoadPlan): Stage[] {
  if (load.stages?.length) return load.stages;
  const result: Stage[] = [];
  const step = load.warmup?.step_s ?? 5;
  if (!saturated(load) && step) {
    for (
      let rate = Math.min(1, load.request_rate);
      rate < load.request_rate;
      rate = Math.min(rate * 2, load.request_rate)
    ) {
      result.push({
        kind: "warmup", duration_s: step,
        request_rate: rate, max_inflight: load.max_inflight,
      });
    }
  }
  result.push({
    kind: "hold", duration_s: load.hold_s ?? 60,
    request_rate: load.request_rate, max_inflight: load.max_inflight,
  });
  return result;
}
export function duration(load: LoadPlan): number {
  return stages(load).reduce((total, stage) => total + stage.duration_s, 0);
}
export function saturated(load: LoadPlan): boolean {
  return load.request_rate === Infinity;
}
export function level(
  stage: Pick<Stage, "request_rate" | "max_inflight">,
): number {
  return stage.request_rate === Infinity
    ? stage.max_inflight
    : stage.request_rate;
}
export function peakLevel(load: LoadPlan): number {
  return Math.max(...stages(load).map(level));
}
export function stageLabel(stage: Stage): string {
  return stage.name ?? `${stage.kind}@${level(stage)}`;
}
export function validateLoadPlan(load: LoadPlan): void {
  if (!load.stages?.length && !(load.request_rate > 0))
    throw new Error("request_rate must be > 0");
  if (!(load.request_rate >= 0)) throw new Error("invalid request_rate");
  if (!Number.isFinite(load.hold_s ?? 60) || (load.hold_s ?? 60) <= 0)
    throw new Error("hold_s must be finite and > 0");
  if (!Number.isFinite(load.warmup?.step_s ?? 5) || (load.warmup?.step_s ?? 5) < 0)
    throw new Error("warmup.step_s must be finite and >= 0");
  for (const stage of [{ ...load, duration_s: load.hold_s ?? 60, kind: "hold" }, ...stages(load)]) {
    if (!Number.isFinite(stage.duration_s) || stage.duration_s <= 0)
      throw new Error("duration_s must be finite and > 0");
    if (!(stage.request_rate >= 0) || stage.request_rate === -Infinity)
      throw new Error("request_rate must be >= 0 or inf");
    if (!Number.isInteger(stage.max_inflight) || stage.max_inflight < 1)
      throw new Error("max_inflight must be an integer >= 1");
    if (stage.kind !== "hold" && stage.kind !== "ramp" && stage.kind !== "warmup")
      throw new Error("stage.kind must be hold, ramp or warmup");
    if ((stage.request_rate === Infinity) !== saturated(load))
      throw new Error("stages cannot switch between finite rate and inf");
  }
  if (
    !Number.isFinite(load.cooldown_timeout_s ?? 180) ||
    (load.cooldown_timeout_s ?? 180) < 0
  )
    throw new Error("invalid cooldown_timeout_s");
  if (load.arrival && !["constant", "poisson"].includes(load.arrival))
    throw new Error("invalid arrival distribution");
  if (
    !Number.isInteger(load.breaker_min_n ?? 20) ||
    (load.breaker_min_n ?? 20) < 1
  )
    throw new Error("invalid breaker_min_n");
  if (
    load.abort_on_error_rate !== undefined &&
    !(load.abort_on_error_rate > 0 && load.abort_on_error_rate <= 1)
  )
    throw new Error("invalid abort_on_error_rate");
}
export function target(load: LoadPlan, t: number): [number, number] {
  let clock = 0,
    rate = load.request_rate,
    concurrency = load.max_inflight;
  for (const s of stages(load)) {
    if (t < clock + s.duration_s) {
      if (s.kind !== "ramp") return [s.request_rate, s.max_inflight];
      const f = Math.max(0, t - clock) / s.duration_s;
      return [
        saturated(load) ? Infinity : rate + (s.request_rate - rate) * f,
        Math.max(
          1,
          Math.floor(concurrency + (s.max_inflight - concurrency) * f),
        ),
      ];
    }
    clock += s.duration_s;
    rate = s.request_rate;
    concurrency = s.max_inflight;
  }
  return [rate, concurrency];
}
export function arrivalTime(load: LoadPlan, volume: number): number {
  let clock = 0,
    previous = load.request_rate;
  for (const s of stages(load)) {
    const start = s.kind === "ramp" ? previous : s.request_rate;
    const slope = (s.request_rate - start) / s.duration_s;
    const mass = ((start + s.request_rate) * s.duration_s) / 2;
    if (volume < mass) {
      if (Math.abs(slope) < 1e-12) return clock + volume / start;
      return (
        clock +
        (volume
          ? (2 * volume) /
            (start + Math.sqrt(Math.max(0, start * start + 2 * slope * volume)))
          : 0)
      );
    }
    volume -= mass;
    clock += s.duration_s;
    previous = s.request_rate;
  }
  return Infinity;
}
export function loadLabel(load: LoadPlan): string {
  return `${saturated(load) ? "concurrency" : "rate"}/${peakLevel(load)}/c${Math.max(load.max_inflight, ...stages(load).map((s) => s.max_inflight))}`;
}
export function resourceLabel(
  resource: import("./model").ResourceProfile,
): string {
  return (
    [
      resource.workers === undefined ? null : `w${resource.workers}`,
      resource.memory,
      resource.cpu ? `cpu${resource.cpu}` : null,
    ]
      .filter(Boolean)
      .join("/") || "default"
  );
}

// Normalized configuration shared by identity hashing and persisted Arm facts.
export function serializeLoadPlan(load: LoadPlan) {
  return {
    ...load,
    request_rate: load.request_rate === Infinity ? "inf" : load.request_rate,
    stages: (load.stages ?? []).map((s) => ({
      ...s,
      request_rate: s.request_rate === Infinity ? "inf" : s.request_rate,
    })),
    arrival: load.arrival ?? "constant",
    seed: load.seed ?? 0,
    hold_s: load.hold_s ?? 60,
    warmup: { step_s: load.warmup?.step_s ?? 5 },
    abort_on_error_rate: load.abort_on_error_rate ?? null,
    breaker_min_n: load.breaker_min_n ?? 20,
    cooldown_timeout_s: load.cooldown_timeout_s ?? 180,
  };
}
